### ISSUE TITLE
SU job assigned too short track

### ISSUE THREAD

**Issue by Mich3lin** [Banjobeni/DerailValley-PersistentJobs#16]
Hello, I've encountered bug, when a 2-drop off job gets assigned single track for both drop-offs which is too short to accommodate the whole consist. This is the job: ![job1](https://github.com/Banjobeni/DerailValley-PersistentJobs/assets/8073586/17df2c4e-4fcc-4360-be74-336e8135dbdc) (**270m**) gets assigned to **G1S** for both parts ![job2](https://github.com/Banjobeni/DerailValley-PersistentJobs/assets/8073586/82f5161d-77a2-4206-8585-b6937eec0e2c) ![job3](https://github.com/Banjobeni/DerailValley-PersistentJobs/assets/8073586/efde075d-9919-42e8-ac90-cd1946195c12) Longer tracks were obviously available at time of accepting job.
![overview](https://github.com/Banjobeni/DerailValley-PersistentJobs/assets/8073586/0183a967-586c-4fef-9785-1c8445fe5252)

**Comment by Banjobeni** [comment]
Hi and thanks for the detailed bug report. That does indeed look like it doesn't fit. Do you have by any chance your Player.log (or Player-prev.log) still available from that session? And/or a savegame from right before accepting this job?

**Comment by Mich3lin** [comment]
I'm not sure if the log file hasn't been overwritten already, but at least I have the save file right before accepting the job. Here it is:
[Derail Valley.zip](https://github.com/Banjobeni/DerailValley-PersistentJobs/files/12882913/Derail.Valley.zip)

**Comment by Banjobeni** [comment]
Try finding the log files anyway, i could search in them for anything specific to that job SU14.

**Comment by Mich3lin** [comment]
Yes, I forgot to mention I've attached both logs already. I've found 
`Matched job and job booklet [HB-SU-14] 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)

[PersistentJobsMod] YardTracksOrganizer.ReserveSpace 147.1m on track HB-G1S. Reserved space is now 187.1m of 259.4m total 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)

Unexpected state: Not enough unreserved space for reservation! Something is not right. 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)

[PersistentJobsMod] YardTracksOrganizer.ReserveSpace 128.7m on track HB-G1S. Reserved space is now 315.8m of 259.4m total 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)`

in the "prev" log.

**Comment by Banjobeni** [comment]
Huh, really interesting. I need to debug.

Actually thanks, I really needed a reason to dig into that part of the code too and probably deep-rewrite it. I took it from the original mod from pre-sim and never touched it. It really looks like it wants to be rewritten, though.

**Comment by Banjobeni** [comment]
I see two bugs in here:
a) Reserving more than what is available
b) Changing a destination track to one that is already part of another dropoff without there really being a need for it (other S tracks are obviously free)

**Comment by Mich3lin** [comment]
One thing I noticed, after doing this job for the third time now, is that during loading I must have coupled the wagons in wrong order. On third try it assigned half to G1 and half to G7, but the wagon numbers were mismatched. I had to shunt the consist back and forth to get the "correct layout". 
I though that if you don't adhere to the SL order, it would sort of recheck the consist and create SU job based on the actual consist layout, than what it was supposed to be like. At least I don't think I ever had this issue in Overhauled, but maybe it's just a coincidence...

**Comment by T3M8029** [comment]
Do we know if this issue is still current?

**Comment by Banjobeni** [comment]
I guess this can still occur. If a multi dropoff job targets specific tracks, we check one by one if we can reserve space on that track. Say for dropoff part 1 we can't reserve space on the intended track A, but we find that B is a suitable replacement, then we change to it and reserve space there. For dropoff part 2 that intended track B, we can't reserve space there but we also can't find a suitable replacement track, we leave it at B. We get two dropoffs to the same track that don't fit there both at the same time.

A better algorithm for this would probably have to handle a couple of iffy edge cases.

**Comment by buck2202** [comment]
Ah, understood re: it being a different issue. I do understand that the intent is for the mod to do its best with track reservation and had observed tracks being _reassigned_, but the "silent failure" of the mod finding **no** suitable track and so issuing the job with an untenable destination was still unexpected to me. I'll think on it a little and maybe open a separate issue for discussion :)

**Comment by buck2202** [comment]
If it's helpful, here's a current-build example of this thread's issue--multipart drop-offs targeting the same track twice with two many cars to fit. Save file is from immediately before accepting the job, the log covers job acceptance. HB-SU-49 targets D1S with blocks of 4 and 5 cars, other S tracks are open, and only one of those sets can fit in D1S

![Image](https://github.com/user-attachments/assets/2fc8e3b0-e081-4521-9113-cbdfd2c9499b)

![Image](https://github.com/user-attachments/assets/ad7e25de-9045-4126-b90b-0af08d695f5c)

![Image](https://github.com/user-attachments/assets/992a7f52-fc8f-46ef-9936-c26f54abae28)

[multipart.zip](https://github.com/user-attachments/files/21808121/multipart.zip)

**Comment by OlympusL** [comment]
> I have the same bug, job assigned whole 13 cars train to a single track as three parts (5-4-4). I don't understand the issue, according to log it assigned different tracks but the job points all three parts to the same track:
> 
> ```
> [PersistentJobsMod] getting random track with free space
> [PersistentJobsMod] 2/5 tracks have at least 74.41456m available
> [PersistentJobsMod] YardTracksOrganizer.ReserveSpace 74.4m on track FF-C9S. Reserved space is now 114.4m of 253.9m total 
> (Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)
> 
> [PersistentJobsMod] getting random track with free space
> [PersistentJobsMod] 2/5 tracks have at least 59.59268m available
> [PersistentJobsMod] YardTracksOrganizer.ReserveSpace 59.6m on track FF-D4S. Reserved space is now 99.6m of 196.8m total 
> (Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)
> 
> [PersistentJobsMod] getting random track with free space
> [PersistentJobsMod] 1/5 tracks have at least 59.59225m available
> [PersistentJobsMod] YardTracksOrganizer.ReserveSpace 59.6m on track FF-D3S. Reserved space is now 99.6m of 196.1m total 
> (Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 39)
> 
> Job FF-SU-15 is assigned to player 
> ```
> 
> I'm attaching job screenshots and Player.log.
> 
> [DV_bug.zip](https://github.com/user-attachments/files/24184101/DV_bug.zip)

If you still have the previous save before accepting the job, you could try my modified version of the mod at #42, which aims to fix an issue where tracks may be mistargeted during the replacement of original tracks with new ones.

**Comment by Banjobeni** [comment]
Hi @Mich3lin, @Ignalion, @OlympusL and @buck2202 . Many thanks for raising and debugging this issue, I greatly appreciate the friendly and helpful discussion.

Sadly I'm not into playing DV anymore, not at the moment. I could adapt and fix code in my mods, but sadly I have zero motivation for investing playtime into DV. I will only release new mod fixes if they have been tested, and since I won't test no updates happen. 

We could break that cycle by you joining my beta testing group and you testing that my code changes work expectedly and don't have side effects. There are already changes in the pipeline that I haven't released, also for the lack of activity in the testing group. So new people who are actively playing and have an eye for wanted and unwanted change are very welcome.

If you are interested, you can find me on discord. I look forward to hearing from you.
